Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -3557,6 +3563,91 @@ def visit_ExprNode(self, node):
     visit_Node = VisitorTransform.recurse_to_children
 
 
+class _ThreadsafetyVisitAssignmentTargetNodes(TreeVisitor):
+    def __call__(self, root):
+        self.coerce_lhs_to_temps = False
+        self.target_info = set()
+        self.visitchildren(root)
+
+        # TODO - in principle lengths of 2 can also be optimized:
+        # with critical sections we can lock 2, and for global/local
+        # scope mutexes we could also lock 2 providing we define a
+        # fixed order.
+        # This potentially helps avoid a few temps.
+        if len(self.target_info) > 1:
+            self.coerce_lhs_to_temps = True
+
+    def visit_Node(self, node):
+        # only go to a depth of one
+        pass
+
+    def _visit_name_or_attribute_node(self, node):
+        if not node.is_target:
+            return
+        if not node.needs_threadsafe_access:
+            return
+        self.target_info.add((
+            node.needs_threadsafe_access, node.threadsafety_data
+        ))
+        if node.needs_threadsafe_access != ExprNodes.ThreadSafetyType.LOCAL:
+            # It's always safe to lock the local scope because it can't
+            # happen recursively.
+            # TODO - when we're properly using critical sections it should
+            # be safe to lock them here too.
+            self.coerce_lhs_to_temps = True
+
+    def visit_NameNode(self, node):
+        self._visit_name_or_attribute_node(node)
+
+    def visit_AttributeNode(self, node):
+        self._visit_name_or_attribute_node(node)
+
+
+class _ThreadsafetyVisitAssignmentNonTargetNodes(VisitorTransform):
+    def __init__(self, coerce_lhs_to_temps, target_info):
+        self.coerce_lhs_to_temps = coerce_lhs_to_temps
+        self.target_info = target_info
+        super().__init__()
+
+    def visit_AssignmentNode(self, node):
+        self.visitchildren(node)
+        return node
+
+    def visit_ExprNode(self, node):
+        # One level only - don't visit children
+        if getattr(self, "is_target", False):
+            return node
+        if not node.is_simple() and self.coerce_lhs_to_temps:
+            return node.coerce_to_temp(None)
+        return node
```

Comment:
So, basically, you want to look only at the immediate RHS expression of assignments here, correct? Can't you do that with `node.rhs`? The place where you use this class already knows that `node` is an `AssignmentNode`.

This visitor class, like the one before, seems unnecessary.